### PR TITLE
Generate Android vectors

### LINKED ISSUE #27: Generate Android vectors for v21 and above
Gradle android build tools 1.4.0-beta1 is bringing support for vectors to png conversion at compile time. We tried it and got back to Victor as it's working better for us. One minor feature requirement I would have though: the build tools are generating pngs only for versions older than v21. For a given vector file (`ic_search.svg`), the output generated folder looks like this:
```
drawable-hdpi
    ic_search.png
drawable-xhdpi
    ic_search.png
drawable-v21
    ic_search.xml
... etc...
```
`ic_search.xml` being an Android vector file.
I don't see why Victor could not do that, anyone has tried to add this feature?

### PR DESCRIPTION
**Pull request by RomainPiel**
_This is not merge-able as it is, I just want to open a discussion._

As an attempt to fix #27 I copied the sources from the Android Studio SVG to Android Vectors converter and made a gradle task based on it: https://github.com/RomainPiel/gradle-svg2android

I modified Victor so that it converts SVGs to Android Vectors instead of PNGs and it's working pretty well with the sample.
SVG2Android hasn't been uploaded to maven central or bintray, I've only played with it with mavenLocal.
So to test this PR:
- clone https://github.com/RomainPiel/gradle-svg2android
- `cd gradle-svg2android && ./gradlew install`

I think a reasonable plan could be to add an option to victor to switch the export format between png and xml. Then while this appears to be stable, we can remove completely `RasterizeTask` and only use `SVG2AndroidTask`.

Limitation:
- I haven't played with a lot of SVGs but it seems that the build tools converter doesn't like `pt` so I converted all of them to `px`

If you think it's useful, either:
- I deploy https://github.com/RomainPiel/gradle-svg2android to bintray
- we integrate the code from https://github.com/RomainPiel/gradle-svg2android into victor

### CODE REVIEW

**Comment by dlew** [comment]
This is awesome. I think we'd want to add an option to convert to PNG vs. Android drawable, just so people can have a safe upgrade path.

One other question - why is gradle-svg2android tied to Gradle vs. being a standalone Java process?

**Comment by RomainPiel** [comment]
Awesome, I have a 7 hours flight tomorrow, I'lltake this time to tidy up the PR.
I made my repo tied to gradle because I wanted to expose a task but you're right, it would be cleaner as a standalone java library.

**Comment by RomainPiel** [comment]
I think it's much more clean now if you want to have a look:
- [x] moved task from romainpiel/svg2android to victor
- [x] deployed romainpiel/svg2android to sonatype
- [x] added a control to enable the export to Android vector drawable. note that densities and default dpi don't really make sense in that case.
- [x] cleaned up as much as possible the logging (i might have missed some lines)

Travis is not passing because the job doesn't do a clean.

File: victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy [hunk 1]
```diff
@@ -83,15 +84,25 @@ class VictorPlugin implements Plugin<Project> {
                     return
                 }
 
-                Task rasterizationTask = project.task("rasterizeSvgsFor${variant.name.capitalize()}", type: RasterizeTask) {
-                    sources = svgFiles
-                    outputDir = project.file("$project.buildDir/generated/res/$flavorName/$buildType.name/svg/")
-                    includeDensities = densities
-                    baseDpi = project.victor.svgDpi
+
+                File conversionOutputDir = project.file("$project.buildDir/generated/res/$flavorName/$buildType.name/svg/")
```
Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy]: I wonder if this should vary based on which type we're outputting - that way, if someone flips `generateVectorDrawables` it won't throw a fit.

File: sample/src/main/svg/ic_laptop.svg [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<svg xmlns="http://www.w3.org/2000/svg" width="24pt" height="24pt" viewBox="0 0 24 24">
+<svg xmlns="http://www.w3.org/2000/svg" width="24px" height="24px" viewBox="0 0 24 24">
```
Comment (dlew) [inline sample/src/main/svg/ic_laptop.svg]: Why the change to `px`? Is the SVG2Android still limited by this?

File: sample/build.gradle [hunk 3]
```diff
@@ -54,6 +54,9 @@ victor {
 
     // Do not generate these densities for SVG assets
     excludeDensities = [ 'ldpi', 'xxxhdpi' ]
+
+    // Do not generate android vector drawables (generate PNGs for the specified densities)
+    generateVectorDrawables = false
```
Comment (dlew) [inline sample/build.gradle]: I'd remove this for now, so that CI passes.

Due to the annoying nature of Gradle plugins, I generally update the sample after the next version goes out.

File: victor/src/main/groovy/com/trello/victor/SVG2AndroidTask.groovy [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    FileCollection sources
+
+    /**
+     * The output directory.
+     */
+    @OutputDirectory
+    File outputDir
+
+    @TaskAction
+    def transform(IncrementalTaskInputs inputs) {
```
Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/SVG2AndroidTask.groovy]: So much of this code is the exact same as the other one - the only part that's different is the actual conversion.

I think this is probably fine for an initial pull, but do you think it'd work to just merge these into a single task that has a generator method provided to it?

File: sample/src/main/svg/ic_laptop.svg (hunk 2 above)
Comment (RomainPiel) [inline sample/src/main/svg/ic_laptop.svg]: Yeah I need to investigate that one. Last time I checked it was throwing an exception. But looking at the source code, there should be some kind of support: https://android.googlesource.com/platform/tools/base/+/master/sdk-common/src/main/java/com/android/ide/common/vectordrawable/VdParser.java#379
I'll try to get to the bottom of this and see if I can create a bug report for the AS project.

File: sample/build.gradle (hunk 3 above)
Comment (RomainPiel) [inline sample/build.gradle]: ok!

File: victor/src/main/groovy/com/trello/victor/SVG2AndroidTask.groovy (hunk 4 above)
Comment (RomainPiel) [inline victor/src/main/groovy/com/trello/victor/SVG2AndroidTask.groovy]: This sounds right, it shouldn't be hard to merge both tasks.

File: victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy (hunk 1 above)
Comment (RomainPiel) [inline victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy]: The problem is that the dexer would complain about duplicate resources. So if we would have to clean both folders all the time from the plugin instead of the tasks.

Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy]: Wait, wouldn't that happen now since we're using the same directory for both tasks? My point is that if we switch based on output, then the user may have both an SVG->PNG and SVG->Android directory at the same time (so they can compare), but both won't be slurped up.

Comment (RomainPiel) [inline victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy]: but if they are under a directory in `build/generated/res/`, they will be slurped up, no?

Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy]: No, the only reason they're picked up is because of `registerResGeneratingTask()`.

Comment (RomainPiel) [inline victor/src/main/groovy/com/trello/victor/VictorPlugin.groovy]: Oh I did not know that!

**Comment by RomainPiel** [comment]
I've merged the tasks and removed the usage of `generateVectorDrawables ` in the sample. I'm still not sure what to do with https://github.com/trello/victor/pull/43#discussion_r60452265

**Comment by RomainPiel** [comment]
regarding https://github.com/trello/victor/pull/43#discussion_r60452328, this is a bug in the Android Studio source code. I've just reported it, so I'll merge it in svg2android as soon as it gets fixed: http://b.android.com/208216

File: victor/src/main/groovy/com/trello/victor/VictorPluginExtension.groovy [hunk 5]
```diff
@@ -42,4 +42,17 @@ class VictorPluginExtension {
      */
     List<String> excludeDensities = []
 
+    /**
+     * [Experimental]
+     *
+     * If set to true, Victor will generate Android vector drawables
+     * If not, it will generate PNGs in all specified densities
```
Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/VictorPluginExtension.groovy]: I think we can merge this now, as long as we specify that it only supports a subset of SVGs (e.g. only `px` values).

File: victor/src/main/groovy/com/trello/victor/RasterizeTask.groovy [hunk 6]
```diff
@@ -84,35 +93,70 @@ class RasterizeTask extends DefaultTask {
...
+    }
+
+    void cleanupInput(@Nullable Density density = null) {
+        File resDir = getResourceDir(density)
+        File file = new File(resDir, getDestinationFile(change.file.name))
+        file.delete()
+    }
+
+    File getResourceDir(@Nullable Density density = null) {
+        String prefix = density? "-${density.name().toLowerCase()}" : ""
```
Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/RasterizeTask.groovy]: Nit - this is technically a suffix.

File: victor/src/main/groovy/com/trello/victor/RasterizeTask.groovy [hunk 7]
```diff
@@ -84,35 +93,70 @@ class RasterizeTask extends DefaultTask {
...
-    File getResourceDir(Density density) {
-        return new File(outputDir, "/drawable-${density.name().toLowerCase()}")
+    void createOutput(@Nullable Density density = null) {
+        File resDir = getResourceDir(density)
+        resDir.mkdirs()
+    }
+
+    void cleanupInput(@Nullable Density density = null) {
+        File resDir = getResourceDir(density)
+        File file = new File(resDir, getDestinationFile(change.file.name))
```
Comment (dlew) [inline victor/src/main/groovy/com/trello/victor/RasterizeTask.groovy]: How does this work? How does it get the `change` variable?

Comment (RomainPiel) [inline victor/src/main/groovy/com/trello/victor/RasterizeTask.groovy]: That's a very good question. What's worrying me is that I'm sure I tested it. I'll triple check again..  Sorry about that.

Comment (RomainPiel) [inline victor/src/main/groovy/com/trello/victor/RasterizeTask.groovy]: Really weird, I checked again and everything works fine and the sample compiles fine. I guess it didn't matter as it was failing in the cleanup. Anyway I fixed it in (yet) another commit.

**Comment by dlew** [comment]
This looks good to me. Let's see what happens with it!
